src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/PooledObjects/PooledArrayBuilder`1.cs

```diff
@@ -2,195 +2,83 @@
+            _count = 0;
         }
-
-        return builder;
-    }
-
-    /// <summary>
-    ///  Retrieves the inner <see cref="_builder"/>.
-    /// </summary>
-    /// <returns>
-    ///  Returns <see langword="true"/> if <see cref="_builder"/> is available; otherwise <see langword="false"/>
-    /// </returns>
-    /// <remarks>
-    ///  This should only be used by methods that will not add to the inner <see cref="_builder"/>.
-    /// </remarks>
-    private readonly bool TryGetBuilder([NotNullWhen(true)] out ImmutableArray<T>.Builder? builder)
-    {
-        builder = _builder;
-        return builder is not null;
     }
 
     /// <summary>
-    ///  Retrieves the inner <see cref="_builder"/> and resets its capacity if necessary.
+    ///  Ensures the inner <see cref="_array"/>'s capacity is at least the specified value.
     /// </summary>
-    /// <returns>
-    ///  Returns <see langword="true"/> if <see cref="_builder"/> is available; otherwise <see langword="false"/>
-    /// </returns>
     /// <remarks>
-    ///  This should only be used by methods that will add to the inner <see cref="_builder"/>.
+    ///  This should only be used by methods that will add to the inner <see cref="_array"/>.
     /// </remarks>
-    private readonly bool TryGetBuilderAndEnsureCapacity([NotNullWhen(true)] out ImmutableArray<T>.Builder? builder)
+    private void EnsureCapacity(int capacity)
     {
-        if (TryGetBuilder(out builder))
+        if (_array.Length < capacity)
         {
-            if (builder.Capacity == 0 && _capacity is int capacity)
-            {
-                builder.Capacity = capacity;
-            }
+            var newArray = ArrayPool<T>.Shared.Rent(capacity);
+            Array.Copy(_array, 0, newArray, 0, Count);
+            ArrayPool<T>.Shared.Return(_array);
+            _array = newArray;
         }
-
-        return builder is not null;
     }
 
-    [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    private void ClearInlineElement(int index)
-    {
-        Debug.Assert(_inlineCount <= InlineCapacity);
-
-        // Clearing out an item makes it potentially available for garbage collection.
-        // Note: On .NET Core, we can be a bit more judicious and only zero-out
-        // fields that contain references to heap-allocated objects.
-
-#if NETCOREAPP
-        if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
-#endif
-        {
-            SetInlineElement(index, default!);
-        }
-    }
-
-    public T this[int index]
+    public readonly T this[int index]
     {
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        readonly get
+        get
         {
-            if (TryGetBuilder(out var builder))
-            {
-                return builder[index];
-            }
-
-            if (index >= _inlineCount)
-            {
-                ThrowIndexOutOfRangeException();
-            }
-
-            return GetInlineElement(index);
+            return _array[index];
```

Comment:
Feels like the indexer should still validate against count, since the array length could be larger and therefore the underlying indexer can't be relied on to throw.